docker: Allow building for non-root user

the-sun-will-rise-tomorrow opened the pull request:
# Motivation

Add options uid, gid, uname, and gname to docker.nix.

Setting these to e.g. 1000, 1000, "user", "user" will build an image which runs and allows using Nix as that user.

# Context

- Fixes https://github.com/nix-community/docker-nixpkgs/issues/62

- Dependencies:
  - [x] https://github.com/NixOS/nixpkgs/pull/271976
  - [x] https://github.com/NixOS/nixpkgs/pull/281520
  - [x] https://github.com/NixOS/nixpkgs/pull/282886

Although (with https://github.com/NixOS/nixpkgs/pull/281520) it's possible to simply do a `chown ... /nix` in `fakeRootCommands`, this is quite inefficient. A change *only* in ownership is not representable as a Docker layer, so the contents of the entire Nix store will end up being copied in the final layer. This is what https://github.com/NixOS/nixpkgs/pull/282886 fixes and this PR builds on.

<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

Ericson2314 commented:
I don't understand everything involved on the docker side, but the code change itself looks sensible good to me. Seems like a good companion to #789 too.

thufschmitt commented:
(marked as draft since it depends on open Nixpkgs PRs, so we don't want to accidentally merge it right now).

I'm a bit reluctant in general when it comes to extending the Docker setup because I'm not sure that we want to maintain a non-trivial Docker setup. But that change is great since it's – mostly – a thin layer on top of the Nixpkgs-side heavy lifting. So :+1:

roberth commented:
Blocked on
- [ ] Backport of https://github.com/NixOS/nix/pull/9854
- [ ] #9546

cole-h commented:
(@roberth FYI you linked to this PR in your "blocked on backport of", but presumably meant to link to one (or all) of the Nixpkgs PRs in the body?)

dhess commented:
Is this ready to merge now?

WxNzEMof commented:
> * [ ]  Backport of [Allow streaming layered containers with non-root Nix store nixpkgs#282886](https://github.com/NixOS/nixpkgs/pull/282886)

Just noticed this was still missing, PR opened: https://github.com/NixOS/nixpkgs/pull/302113

WxNzEMof commented:
Rebased; I think all dependencies are now satisfied.

mmlb commented:
Looks like this has fallen through the cracks?

dhess commented:
Yes, can we get this merged, please?

roberth reviewed (changes requested):
This PR still needs:
- tests: new test case in `tests/nixos/`?
- documentation, `doc/manual/source/installation/installing-docker.md`

I apologize for the other maintainers who have merged insufficiently documented and untested code. That should not have happened.

the-sun-will-rise-tomorrow commented:
> tests: new test case in `tests/nixos/`?

Added a very simple test. (I tried to put together a more complete test which does some store operations, but ran into [this issue](https://github.com/NixOS/nix/issues/6835).)

> documentation, `doc/manual/source/installation/installing-docker.md`

Thanks, added.

the-sun-will-rise-tomorrow commented:
> Added a very simple test. (I tried to put together a more complete test which does some store operations, but ran into [this issue](https://github.com/NixOS/nix/issues/6835).)

I managed to get a more thorough test working with the help of a binary cache server machine, but I still can't figure out how to write a test that tries to actually build a derivation inside the container. We need to provide all dependencies, but for some reason it's not happy with what's on the cache and tries to download some bootstrap tarballs, even if I try to warm up the cache by doing the same thing there first.

roberth commented:
Nix's normal logging isn't the most helpful in this case because it gives you all the paths it needs, but if you pas `-v` or `-vv` it will tell you more about what it actually does. Probably the first path it queries is the one you need to provide (and repeat that solution strategy until it works).
What may also help is to build a simple derivation with `builtins.derivation` so that you control accurately what it depends on.
Otherwise you'll be copying Nixpkgs into the test derivation, slowing down the test.

I don't see the cache you're referring to, but I think you could bake the dependencies into the image.
For the expression file you put in the image, you could use a function `storepath: builtins.appendContext storepath { ${storepath} = { path = true; } }` to convert plain strings containing store paths into store paths that the Nix inside the image understands.

Feel free to push a work in progress commit so I can have a look.

the-sun-will-rise-tomorrow commented:
> What may also help is to build a simple derivation with `builtins.derivation` so that you control accurately what it depends on.

Great idea, it worked, thanks!

roberth reviewed (approved):
This is great!

dhess commented:
Thanks for this, it'll be great to run Nix in containers as non-root.

If I understand correctly (from https://github.com/NixOS/nix/pull/5380#issue-1024792160), pretty much the only thing that won't work is finding GC roots; is that right?

the-sun-will-rise-tomorrow commented:
> If I understand correctly (from [#5380 (comment)](https://github.com/NixOS/nix/pull/5380#issue-1024792160)), pretty much the only thing that won't work is finding GC roots; is that right?

That would apply to a daemon (multi-user) setup. This container has a single-user installation, which makes sense for OCI containers. I think GC should work fine if everything in the container runs as the same user.

roberth commented:
The container has a proper store db for the [store object] parts that aren't the path or data, like the references for GC. Single user setups are supported. The only problems I would expect is to perhaps corrupting the store by accident because it is less protected. Also the sandbox may or may not be possible to isolate properly, depending on the container runtime and settings.

[store object]: https://nix.dev/manual/nix/latest/glossary.html#gloss-store-object

dhess commented:
Thanks all!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-11-18-nix-team-meeting-minutes-195/56229/1
